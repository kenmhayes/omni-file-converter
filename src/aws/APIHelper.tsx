import { API } from 'aws-amplify';

const SESSION_API_NAME = 'SessionGatewayAPI';

export interface CreateSessionRequest {
  conversionRequests: FileConversionInput[];
}

export interface FileConversionInput {
  fileName: string;
  originalFileFormat: string;
  convertedFileFormat: string;
}

interface CreateSessionResponse {
  sessionId: string;
  fileNameToId: { [key: string]: string };
}

export interface GetSessionResponse {
  expirationTime: number;
  conversions: ConversionOutput[];
}

export interface ConversionOutput {
  id: string;
  fileName: string;
  conversionStatus: string;
  originalFileFormat: string;
  convertedFileFormat: string;
}

// eslint-disable-next-line max-len
export const createSession = async (request: CreateSessionRequest): Promise<CreateSessionResponse> => API.post(
  SESSION_API_NAME,
  '/createsession',
  {
    body: request,
  },
);

export const getSession = async (sessionId: string): Promise<GetSessionResponse> => API.get(
  SESSION_API_NAME,
  '/getsession',
  {
    queryStringParameters: {
      sessionid: sessionId,
    },
  },
);
