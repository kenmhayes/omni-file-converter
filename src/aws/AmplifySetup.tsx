import { Amplify, Auth, Storage } from 'aws-amplify';

function AmplifySetup() {
  /**
 * Initialize Cognito authentication
 */
  Amplify.configure({
    Auth: {
      identityPoolId: process.env.REACT_APP_IDENTITY_POOL_ID,
      region: process.env.REACT_APP_AWS_REGION,
    },
    Storage: {
      AWSS3: {
        bucket: process.env.REACT_APP_S3_BUCKET,
        region: process.env.REACT_APP_AWS_REGION,
      },
    },
  });

  Auth.configure();

  const getCredentials = async () => {
    await Auth.currentCredentials();
  };

  getCredentials();

  // Configure S3 storage
  Storage.configure({
    customPrefix: {
      public: '',
    },
  });
}

export default AmplifySetup;
