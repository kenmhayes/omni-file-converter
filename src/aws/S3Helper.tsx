import { Storage } from 'aws-amplify';

export const getS3ObjectUrl = async (objectKey: string): Promise<string> => Storage.get(
  `${process.env.REACT_APP_S3_CONVERTED_PREFIX}${objectKey}`,
);

export const putS3Object = async (
  objectKey: string,
  convertType: string,
  file: File,
): Promise<void> => {
  await Storage.put(`${process.env.REACT_APP_S3_ORIGINAL_PREFIX}${convertType}/${objectKey}`, file);
};
