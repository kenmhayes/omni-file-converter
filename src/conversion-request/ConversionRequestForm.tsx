import React, { useState, useEffect } from 'react';
import { Button } from 'react-bootstrap';
import Select, { SelectOption } from './Select';
import {
  FILE_TYPE_CONVERSION_MAP, FileType, DEFAULT_FILE_TYPE, FileTypeConversionMetadata,
} from '../constants/FileTypes';
import FileUpload from './FileUpload';
import { putS3Object } from '../aws/S3Helper';

function createSelectOptionFromFileType(fileType: FileType): SelectOption {
  return { id: fileType.key, displayValue: fileType.displayValue };
}

function getFileTypeFromMap(fileTypeKey: string): FileType {
  const metadata: FileTypeConversionMetadata = FILE_TYPE_CONVERSION_MAP[fileTypeKey];
  if (metadata) {
    return metadata.fileType;
  }
  return DEFAULT_FILE_TYPE;
}

/**
 * Form for creating a conversion request, such as indicating file types and files to upload
 * @returns
 *   a react element
 */
function ConversionRequestForm() {
  const [originalType, setOriginalType] = useState<FileType>(DEFAULT_FILE_TYPE);
  const [convertType, setConvertType] = useState<FileType>(DEFAULT_FILE_TYPE);
  const [convertTypeSelectOptions, setConvertTypeSelectOptions] = useState<SelectOption[]>([]);
  const [files, setFiles] = useState<File[]>([]);

  const originalFileTypeSelectOptions: SelectOption[] = Object.keys(FILE_TYPE_CONVERSION_MAP)
    .map((fileTypeKey: string) => createSelectOptionFromFileType(getFileTypeFromMap(fileTypeKey)));

  const originalTypeSelectionHandler = (selectedFileType: string) => {
    setOriginalType(getFileTypeFromMap(selectedFileType));
  };
  const convertTypeSelectionHandler = (selectedFileType: string) => {
    setConvertType(getFileTypeFromMap(selectedFileType));
  };

  const onUploadClick = async () => {
    // Temp code, so untested
    files.forEach(async (file: File) => {
      await putS3Object(file.name, file);
    });
  };

  useEffect(() => {
    setConvertType(DEFAULT_FILE_TYPE);
    setFiles([]);
    if (originalType.key !== DEFAULT_FILE_TYPE.key) {
      const conversionData: FileTypeConversionMetadata = FILE_TYPE_CONVERSION_MAP[originalType.key];
      setConvertTypeSelectOptions(
        conversionData.supportedConversionTypeKeys.map(
          (key: string) => createSelectOptionFromFileType(getFileTypeFromMap(key)),
        ),
      );
    } else {
      setConvertTypeSelectOptions([]);
    }
  }, [originalType]);

  return (
    <div>
      <Select
        options={originalFileTypeSelectOptions}
        onSelectionChange={originalTypeSelectionHandler}
        placeholderText="Original file type"
      />
      <div>{originalType.displayValue}</div>
      <Select
        options={convertTypeSelectOptions}
        onSelectionChange={convertTypeSelectionHandler}
        placeholderText="Converted file type"
      />
      <div>{convertType.displayValue}</div>
      <FileUpload fileTypes={originalType.extensions} files={files} setFiles={setFiles} />
      <Button onClick={onUploadClick}>Upload</Button>
    </div>
  );
}

export default ConversionRequestForm;
