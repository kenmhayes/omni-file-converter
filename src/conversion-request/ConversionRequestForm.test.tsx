import React from 'react';
import { create } from 'react-test-renderer';
import { fireEvent, render, waitFor } from '@testing-library/react';
import ConversionRequestForm from './ConversionRequestForm';
import { MemoryRouter } from 'react-router-dom';
import * as APIHelper from '../aws/APIHelper';
import * as S3Helper from '../aws/S3Helper';

const apiSpy = jest.spyOn(APIHelper, 'createSession');
const s3Spy = jest.spyOn(S3Helper, 'putS3Object');

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
   ...jest.requireActual('react-router-dom') as any,
  useNavigate: () => mockedUsedNavigate,
}));

it('renders a conversion request form', () => {
  const component = create(
    <MemoryRouter>
      <ConversionRequestForm/>
    </MemoryRouter>
    
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('empties files on file type change', () => {
  const component = <MemoryRouter>
    <ConversionRequestForm/>
  </MemoryRouter>;
  const newFile = new File(['hello'], 'hello.jpg', {type: 'image/jpg'})

  const {queryByText, getByDisplayValue } = render(component);

  // Select jpg in the first select
  fireEvent.change(getByDisplayValue("Original file type"), { target: { value: 'jpg'}});

  // Find file input, doesn't seem possible with react testing library with this 3rd party component
  const fileInputQuery = document.querySelector("input");
  expect(fileInputQuery).toBeTruthy();
  const fileInput = fileInputQuery as HTMLInputElement;

  fireEvent.change(fileInput, { target: { files: [newFile] }});
  expect(queryByText(newFile.name)).toBeTruthy();
  
  // Change original file type
  fireEvent.change(getByDisplayValue('jpeg/jpg'), { target: { value: 'png'}});

  expect(queryByText(newFile.name)).toBeFalsy();
});

it('calls API on upload click', async () => {
  apiSpy.mockResolvedValue({
    sessionId: "Session Id",
    fileNameToId: {
      "hello.jpg": "File Id"
    }
});
  const newFile = new File(['hello'], 'hello.jpg', {type: 'image/jpg'})

  const component = <MemoryRouter>
    <ConversionRequestForm/>
  </MemoryRouter>;

  const {getByRole, queryByRole, getByDisplayValue, findByText } = render(component);

  // Select types
  fireEvent.change(getByDisplayValue("Original file type"), { target: { value: 'jpg'}});
  fireEvent.change(getByDisplayValue("Converted file type"), { target: { value: 'png'}});

  // Find file input, doesn't seem possible with react testing library with this 3rd party component
  const fileInputQuery = document.querySelector("input");
  expect(fileInputQuery).toBeTruthy();
  const fileInput = fileInputQuery as HTMLInputElement;

  fireEvent.change(fileInput, { target: { files: [newFile] }});

  const uploadButtonTextOption = { name: "Upload" };
  expect(queryByRole("button", uploadButtonTextOption)).toBeTruthy();

  const uploadButton = getByRole("button", uploadButtonTextOption);
  fireEvent.click(uploadButton);

  await waitFor(async () => expect(apiSpy).toHaveBeenCalledWith({
    conversionRequests: [
      {
        fileName: "hello.jpg",
        originalFileFormat: "jpg",
        convertedFileFormat: "png"
      }
    ]
  }));
  await waitFor(async () => expect(s3Spy).toHaveBeenCalledWith("File Id.jpg", "png", newFile));

  expect(mockedUsedNavigate).toHaveBeenCalledWith("/Session Id");
});
