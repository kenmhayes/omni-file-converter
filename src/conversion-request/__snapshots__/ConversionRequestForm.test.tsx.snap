// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a conversion request form 1`] = `
<div>
  <div
    className="mt-4 card"
  >
    <div
      className="card-header"
    >
      <h4>
        Step #1: Choose file formats
      </h4>
    </div>
    <div
      className="card-body"
    >
      <div
        className="container"
      >
        <div
          className="row pb-4"
        >
          <div
            className="col-3"
          >
            <select
              className="form-select"
              onChange={[Function]}
            >
              <option
                value=""
              >
                Original file type
              </option>
            </select>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="col-3"
          >
            <select
              className="form-select"
              onChange={[Function]}
            >
              <option
                value=""
              >
                Converted file type
              </option>
            </select>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="mt-4 card"
  >
    <div
      className="card-header"
    >
      <h4>
        Step #2: Choose files
      </h4>
    </div>
    <div
      className="card-body"
    >
      <div
        className="container"
      >
        <div
          className="row pb-4"
        >
          <label
            className="sc-aXZVg guZdik  "
            htmlFor="file"
            onClick={[Function]}
          >
            <input
              accept=""
              disabled={false}
              name="file"
              onChange={[Function]}
              onClick={[Function]}
              type="file"
            />
            <svg
              fill="none"
              height="32"
              viewBox="0 0 32 32"
              width="32"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.33317 6.66667H22.6665V16H25.3332V6.66667C25.3332 5.196 24.1372 4 22.6665 4H5.33317C3.8625 4 2.6665 5.196 2.6665 6.66667V22.6667C2.6665 24.1373 3.8625 25.3333 5.33317 25.3333H15.9998V22.6667H5.33317V6.66667Z"
                fill="#0658C2"
              />
              <path
                d="M10.6665 14.6667L6.6665 20H21.3332L15.9998 12L11.9998 17.3333L10.6665 14.6667Z"
                fill="#0658C2"
              />
              <path
                d="M25.3332 18.6667H22.6665V22.6667H18.6665V25.3333H22.6665V29.3333H25.3332V25.3333H29.3332V22.6667H25.3332V18.6667Z"
                fill="#0658C2"
              />
            </svg>
            <div
              className="sc-eqUAAy jWkLDY"
            >
              <span
                className="sc-fqkvVR kFhUBM"
              >
                <span>
                  Upload
                </span>
                 or drop a file right here
              </span>
              <span
                className="file-types"
                title="types: "
              />
            </div>
          </label>
        </div>
        <div
          className="row"
        >
          <div
            className="container"
          >
            <h6>
              Uploaded files:
            </h6>
            <div
              className="list-group list-group-flush"
              onKeyDown={[Function]}
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="mt-4 card"
  >
    <div
      className="card-header"
    >
      <h4>
        Step #3: Upload your files
      </h4>
    </div>
    <div
      className="card-body"
    >
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Upload
      </button>
    </div>
  </div>
</div>
`;
